src/finch/autoschedule/compiler.py

```diff
@@ -136,52 +88,247 @@ class LogicLowerer:
+                    (
+                        *[ntn.Assign(k, v) for k, v in dim_size_vars.items()],
+                        ntn.Assign(agg_res, declaration),
+                        body,
+                        ntn.Assign(agg_res, ntn.Freeze(agg_res, ntn.Literal(op))),
+                    )
+                )
+
+            case Plan((Produces(args),)):
+                assert len(args) == 1, "Only single return object is supported now"
+                match args[0]:
+                    case Reorder(Relabel(Alias(name), idxs_1), idxs_2) if set(
+                        idxs_1
+                    ) == set(idxs_2):
+                        raise NotImplementedError("TODO: not supported")
+                    case Reorder(Alias(name) as alias, _) | Relabel(
+                        Alias(name) as alias, _
+                    ):
+                        tbl_var = table_vars[alias]
+                    case Alias(name) as alias:
+                        tbl_var = table_vars[alias]
+                    case any:
+                        raise Exception(f"Unrecognized logic: {any}")
+                return ntn.Return(tbl_var)
+
+            case Plan(bodies):
+                func_block = ntn.Block(
+                    tuple(self(body, table_vars, dim_size_vars) for body in bodies)
+                )
+                return ntn.Module(
+                    (
+                        ntn.Function(
+                            ntn.Variable("func", np.ndarray),
+                            tuple(var for var in table_vars.values()),
+                            func_block,
+                        ),
+                    )
+                )
 
             case _:
                 raise Exception(f"Unrecognized logic: {ex}")
 
 
+def format_queries(ex: LogicNode) -> LogicNode:
+    return _format_queries(ex, bindings={})
+
+
+def _format_queries(ex: LogicNode, bindings: dict) -> LogicNode:
+    # TODO: continue rep_construct & SuitableRep implementation
+    def rep_construct(a):
+        return a
+
+    class SuitableRep:
+        def __init__(self, bindings):
+            pass
+
+        def __call__(self, obj):
+            return np.ndarray
+
+    match ex:
+        case Plan(bodies):
+            return Plan(tuple(_format_queries(body, bindings) for body in bodies))
+        case Query(lhs, rhs) if not isinstance(rhs, Reformat | Table):
+            assert isinstance(rhs, LogicExpression)
+            rep = SuitableRep(bindings)(rhs)
+            bindings[lhs] = rep
+            tns = rep_construct(rep)
+            return Query(lhs, Reformat(tns, rhs))
+        case Query(lhs, rhs) as query:
+            bindings[lhs] = SuitableRep(bindings)(rhs)
+            return query
+        case _:
+            return ex
+
+
+def initwrite(*args):  # TODO: figure out the implementation
+    raise NotImplementedError
+
+
+# TODO: replace with appropriate Tensor class
+_TensorType = np.ndarray
+
+
+def record_tables(
```

Comment:
This is essentially what "defer_tables" was in Finch.jl. I'm glad for the rename and redesign here.